File: 1563-stone-game-v/Solution.java
class Solution {

    public int stoneGameV(int[] stoneValue) {
        int n = stoneValue.length;

        // Prefix sum array
        int[] prefix = new int[n + 1];

        for (int i = 0; i < n; i++) {
            prefix[i + 1] = prefix[i] + stoneValue[i];
        }

        /*
         * dp[l][r] = maximum score Alice can obtain
         * from the subarray stoneValue[l...r].
         */
        int[][] dp = new int[n][n];

        /*
         * Process intervals from smaller to larger.
         * An interval of length 1 has score 0 because
         * it cannot be split.
         */
        for (int len = 2; len <= n; len++) {

            for (int l = 0; l + len <= n; l++) {

                int r = l + len - 1;

                // Try every possible split point.
                for (int k = l; k < r; k++) {

                    // Sum of stoneValue[l...k]
                    int leftSum = prefix[k + 1] - prefix[l];

                    // Sum of stoneValue[k+1...r]
                    int rightSum = prefix[r + 1] - prefix[k + 1];

                    if (leftSum < rightSum) {

                        /*
                         * Right part has the larger sum,
                         * so Bob removes it.
                         *
                         * Alice keeps the left part and
                         * receives leftSum.
                         */
                        dp[l][r] = Math.max(
                            dp[l][r],
                            leftSum + dp[l][k]
                        );

                    } else if (leftSum > rightSum) {

                        /*
                         * Left part has the larger sum,
                         * so Bob removes it.
                         *
                         * Alice keeps the right part and
                         * receives rightSum.
                         */
                        dp[l][r] = Math.max(
                            dp[l][r],
                            rightSum + dp[k + 1][r]
                        );

                    } else {

                        /*
                         * Both parts have equal sums.
                         *
                         * Alice can choose which part to keep,
                         * so choose the one with the better
                         * future score.
                         */
                        dp[l][r] = Math.max(
                            dp[l][r],
                            leftSum + Math.max(
                                dp[l][k],
                                dp[k + 1][r]
                            )
                        );
                    }
                }
            }
        }

        return dp[0][n - 1];
    }
}

File: 1563-stone-game-v/Solution.py
from typing import List


class Solution:

    def stoneGameV(self, stoneValue: List[int]) -> int:
        n = len(stoneValue)

        # Prefix sum array
        prefix = [0] * (n + 1)

        for i in range(n):
            prefix[i + 1] = prefix[i] + stoneValue[i]

        # dp[l][r] = maximum score Alice can obtain
        # from the subarray stoneValue[l...r]
        dp = [[0] * n for _ in range(n)]

        # Process intervals from smaller to larger
        for length in range(2, n + 1):

            for l in range(n - length + 1):
                r = l + length - 1

                # Try every possible split
                for k in range(l, r):

                    # Sum of stoneValue[l...k]
                    left_sum = prefix[k + 1] - prefix[l]

                    # Sum of stoneValue[k+1...r]
                    right_sum = prefix[r + 1] - prefix[k + 1]

                    if left_sum < right_sum:

                        # Bob removes the right part.
                        # Alice keeps the left part.
                        dp[l][r] = max(
                            dp[l][r],
                            left_sum + dp[l][k]
                        )

                    elif left_sum > right_sum:

                        # Bob removes the left part.
                        # Alice keeps the right part.
                        dp[l][r] = max(
                            dp[l][r],
                            right_sum + dp[k + 1][r]
                        )

                    else:

                        # Equal sums:
                        # Alice can choose either side.
                        dp[l][r] = max(
                            dp[l][r],
                            left_sum + max(
                                dp[l][k],
                                dp[k + 1][r]
                            )
                        )

        return dp[0][n - 1]
